### ISSUE TITLE
Identical directions in Test Results on CHCP Dataset

### ISSUE THREAD

**Issue by tensor2023** [AxelElaldi/e3so3_conv#3]
I really appreciate all the hard work that’s gone into this project!
I’ve run into an issue while training on the Chinese-HCP dataset. The test results are off—every voxel has the same direction. I’ve attached both the test output and the loss curve for reference.
Also, just to note, the "epoch" label in the loss curve title is actually meant to be "step"—that was my mistake.
Any advice would be great!
![1](https://github.com/user-attachments/assets/2793931c-8748-4b7e-a63c-a194655c675f)
![image](https://github.com/user-attachments/assets/a94a8bac-87cf-4c85-9ee6-51138330e860)
![image](https://github.com/user-attachments/assets/d7bd7a12-bb4f-44eb-963d-ac3a9e784158)

**Comment by tensor2023** [comment]
Good news—I’ve already fixed the issue! I realized the problem was that I had set concatenate to True while keeping the kernel_sizeSpa at 3, which caused a lot of zero-padding during convolution, making it hard for the network to learn. Then I ran into another issue: the non-negativity constraint wasn’t working, and the model barely learned anything because the 0th-order spherical harmonics of x_equi kept showing a lot of negative values. Thankfully, I fixed that too with torch.clamp().

**Comment by neel-dey** [comment]
Hi @tensor2023,

It's great that your problem got solved! While clamping can also work, I'm not sure why the non-neg regularization isn't working in your case. Can you post a training call / script? Thanks.

**Comment by tensor2023** [comment]
Without using .clamp(0, None), the output of torch.sum(x_equi[:, :, 0:1], axis=1, keepdim=True) shows many negative values, going as low as -1000, making training ineffective. My response function is calculated after b0 normalization, so these large negative values are unexpected.I have not made any other changes to the code.
Additionally, I noticed that whenever x_equi[:, :, 0:1], axis=1, keepdim=True) produces negative values, the maximum value of x_reconstructed exceeds 1, reaching around 20, while the minimum also becomes negative. This seems to be causing oscillations in the loss function.
I've already changed 30 random seeds, and the results are still the same.
![image](https://github.com/user-attachments/assets/d1b10c17-7cea-433b-9cbc-e4a17036dfab)

**Comment by tensor2023** [comment]
Additionally, I noticed that whenever torch.sum(x_equi[:, :, 0:1], axis=1, keepdim=True) produces negative values, the maximum value of x_reconstructed exceeds 1, reaching around 20, while the minimum also becomes negative. This seems to be causing oscillations in the loss function.

**Comment by neel-dey** [comment]
Thanks for sharing. As above, if you could post your training script or call that would help us debug.

Just going off of the loss plots, this is not a seed issue, it appears to be standard unstable network training. It's likely that the default learning rate in the repo is too high for your dataset, could you please try `--lr 1e-3` or `--lr 3e-4` and tell us if that makes it stable?

**Comment by tensor2023** [comment]
Thank you for responding and for taking a look at my code. I've attached my code with only changes to the data input and optimizer.
[RT-ESD_reproduction_code.zip](https://github.com/user-attachments/files/17535491/RT-ESD_reproduction_code.zip)

And I noticed if the following step is removed and replaced with a neural network, the training becomes stable, with results similar to those described in the paper：

```python
if self.n_equi != 0:
    x_deconvolved_equi_shc = self.get_equi_sh(x_deconvolved_equi)  # B x out_channels_equi x C x X x Y x Z
else:
    x_deconvolved_equi_shc = None
```

The issue seems to be with this line:

```python
x = torch.einsum('ijklmn,kp->ijplmn', x, self.S2SH)
```

where `self.S2SH` has a shape of `[3072, 190]`, and `self.S2SH[:,0]` contains some negative values, causing `x` to have many negative values and leading to unstable training. I don’t understand why projecting from an isotropic spherical function to the 0th-order spherical harmonics would result in negative values, as the 0th-order spherical harmonics are theoretically always positive, so the projection values should be positive as well. I tried adjusting all hyperparameters, including setting the learning rate to `1e-5`, ran experiments on different GPUs (RTX 3090 and GTX 1080), switched to the fibercup dataset, and tested every 10 epochs, but the results remain more or less the same.
![image](https://github.com/user-attachments/assets/4f0d734a-1959-4cb2-8fc2-a636100da5b4)
![image](https://github.com/user-attachments/assets/82dcca3e-bf92-42a3-bb6a-4c0ab2c4a8c9)
I would greatly appreciate your insights or any suggestions you may have. Thank you once again for your time and support!

**Comment by tensor2023** [comment]
I have resolved the issue related to negative values by adjusting the code as follows:

```python
import numpy as np
from dipy.core.sphere import Sphere
from dipy.reconst.shm import sh_to_sf_matrix

vector = np.load('/home/xqgao/D/re/e3so3_conv-main/utils/vector3072_3.npy').astype(np.float32)

sphere_bvecs = Sphere(xyz=vector)

sh_order = 18

design_matrix = sh_to_sf_matrix(sphere_bvecs, sh_order)
np.save('/home/xqgao/D/re/e3so3_conv-main/utils/spatial2spectral_3072190.npy', design_matrix[1])
np.save('/home/xqgao/D/re/e3so3_conv-main/utils/spectral2spatial_1903072.npy', design_matrix[0])
```

After making this adjustment, my training process has become much more stable. However, it seems that the network is still not learning effectively. I would greatly appreciate any further guidance or suggestions from the community to help improve the model's performance. Your support is highly valued, and I look forward to any insights you may have.

Thank you in advance!
![image](https://github.com/user-attachments/assets/add42075-6bd0-4efb-8f72-871907a355aa)
![image](https://github.com/user-attachments/assets/c9dd7a6e-08e2-4dd7-99e3-c9f05e994892)

**Comment by tensor2023** [comment]
Thank you for the support! After reducing the learning rate, the issue has been resolved, and everything is back to normal.

**Comment by neel-dey** [comment]
Hi @tensor2023, very sorry we dropped the ball on responding to this. 

Glad you found the issue! What was the required learning rate for this dataset, btw?

Also, you might be interested in our new work on this topic at this year's NeurIPS conference:
paper: https://arxiv.org/abs/2411.11819
code: https://github.com/AxelElaldi/fast-equivariant-deconv 

It's the same general approach as RT-ESD, but quite a bit faster to train and more spatially regularized.

**Comment by tensor2023** [comment]
Hi @neel-dey , thank you for sharing your exciting work! Actually, I applied a learning rate warmup, peaking at 0.0017, and then decaying. I believe the main issue lies in the computations of the spatial2spectral and spectral2spatial, rather than the learning rate. The high condition number of the tensors caused anomalies during matrix inversion. By using the approach I mentioned earlier to replace the inversion step and training for a sufficient duration, the results improved significantly. And the figure shows a part of my reproduced results.

I’ve read your NeurIPS paper and was impressed by the hemispherical graph convolution, especially its robustness to low-resolution inputs. The fast version and its efficiency are truly appealing!
![image](https://github.com/user-attachments/assets/92a8dbc0-9499-489e-9540-37fac8607a0b)
